user/super/com/google/gwt/emul/java/util/List.java

```diff
@@ -34,6 +35,71 @@
 @JsType
 public interface List<E> extends Collection<E> {
 
+  @JsIgnore
+  static <E> List<E> of() {
+    return Collections.unmodifiableList(Collections.emptyList());
+  }
+
+  @JsIgnore
+  static <E> List<E> of(E e1) {
```

Comment:
Are these types of methods really needed, given that the emulation does not optimize to avoid creating an array? Do we get these methods for free if we only emulate `of(E... elements)` because JDT will match any other call to that single emulated method then?

Reply:
You're almost certainly right that this would work. My hesitancy is around allowing the compiler to _not_ inline the varargs version of this method. Consider 10 calls to `of(E)` - we would rather a single `function of(e1)` implementation that creates the array literal, the array, and then calls `of(arr)` so that each callsite is simply `a(b)`, rather than 10 callsites that are roughly `a(c(d(e, 1), e, 123, 0, b))` (the required boilerplate to create an array with a class literal, and the element `b`, wrapped in a call to `a(..)`).

Reply:
Hmm I see, interesting point. Should we then communicate this intent in code by adding `@DoNotInline`? Possibly with a comment that this could be removed on some methods if someone provides a specialized implementation for the first couple methods that avoids array creation?

Also I note that all the `of()` methods create an array which is then immediately cloned again in `of(E... )` because that API method can not trust its input and needs to make a defensive, unmodifiable copy. But the array created in all the `of()` factory methods can be used directly as it is known to be not referenced by user code. So maybe we can change the code a bit to

```java
static <E> List<E> of(E e1, E e2, E e3) {
  return ofInternal(new Object[] {e1, e2, e3});
}

@JsIgnore
static <E> List<E> of(E... elements) {
  return ofInternal(ArrayHelper.unsafeClone(elements, 0, elements.length));
}

@SuppressWarnings("unchecked")
private static <E> List<E> ofInternal(Object[] elements) {
  for (int i = 0; i < elements.length; i++) {
    checkNotNull(elements[i]);
  }
  return Collections.unmodifiableList(Arrays.asList((E[]) elements));
}
```

Now the `null` check is later, but I assume that it will only trigger in rare cases anyways as JavaDoc states that `null` elements are forbidden. So I would be fine if in these rare cases an array has been unnecessarily cloned beforehand.

Reply:
I would assume (but do not know) that the compiler won't just inline without weighing tradeoffs (and for gwt, closure-compiler, almost all tradeoffs are based on output size).

Good catch on the copy (or rather, our absence of a defensive copy), that is definitely unsafe. Unfortunately, we can't have private methods in interfaces while we still support sourceLevel=8, so make our own static method, with `__` prefix or something to make it unlikely to collide? Make a new package-protected class in java.util for this? Other ideas I am not noticing?

Reply:
> Good catch on the copy (or rather, our absence of a defensive copy), that is definitely unsafe.

Hmm maybe you misread: I said that the code does too many defensive copies right now. The object array created internally by all the `of(e1, e2, e3, ..)` methods should not be cloned again in `of(E... elements)`, because we already know that these arrays are effectively immutable. But because `of(E... elements)` is public API and needs to make a defensive copy, so we need an internal method to delegate to.

> Make a new package-protected class in java.util for this?

I would say so. Something like CollectionsHelper. I would put it in https://github.com/gwtproject/gwt/tree/main/user/super/com/google/gwt/emul/javaemul/internal 

If someone wants to optimize the 1-3 parameter factory methods, that helper class could also hold the specialized list/set/map implementations needed.

Reply:
I understood what you said, but mistakenly believed that the current patch had avoided a copy (though of course unsafeClone does clone the array, just in a way that loses the ctm, etc).

I don't love javaemul as a place to put this, since those types are _almost_ only about how to handle mapping js<->java, but it does seem as good of a place as any to handle this. 

ICYMI, the initial review rejected even the 1 param case as having its own implementation.

Reply:
> ICYMI, the initial review rejected even the 1 param case as having its own implementation.

Puh 2019, hehe. But sure he also has a point of not introducing too many implementations  if you just use List.of variants in a small application.

Reply:
> 2019

Yeah, this patch has been backed up behind a long run of other patches to get us building properly on Java >8.

As far as size goes, I think it is very reasonable to keep out other implementations - each class costs something extra, and each List will need its own corresponding Iterator type (or you squander any performance gains you happen to make). The benefits of a specialized type are likely very slim indeed - especially since we're focusing on the use case of needing to iterate a collection that by definition can't have more than 10 items - are you going to be iterating collections of collections in a tight loop? If that's the case... its probably time for some specialized code in your own app to address that performance issue.

Thanks for taking a look, nice to async chat with you again. I'll get this updated in the next day or so.